Port to use crossbeam instead of mpsc_select

Munksgaard opened the pull request:
Fixes #45 

Todo:
 - [x] hselect
 - [x] iselect
 - [x] ChanSelect
 - [x] offer!
 - [x] try_offer!
 - [x] chan_select!

laumann commented on examples/hselect.rs:
```diff
@@ -0,0 +1,28 @@
+
+    let receivers = vec!(rcs, rcu);
+
+    let () = tcs.send("Hello, World from TCS!".to_string()).close();
```
This pattern looks odd to me, `let () = …`. I thought the idiomatic way to discard results is `let _ = …`. In this case, I don't think calls to `.close()` have `#[must_use]`

laumann commented on src/lib.rs:
```diff
@@ -134,6 +131,24 @@ pub struct Rec<P>(PhantomData<P>);
 /// out of.
 pub struct Var<N>(PhantomData<N>);
 
+trait HasReceiver {
```
Why is this necessary? To allow pulling the `Receiver<_>` out of different channels types?

EDIT Doesn't appear to be used.

laumann commented on src/lib.rs:
```diff
@@ -402,36 +416,22 @@ pub fn hselect<E, P, A>(
-        for handle in handles.iter_mut() {
-            // Clean up
-            unsafe {
-                handle.remove();
```
It's gonna be nice to get rid of this :+1: It was always a bit yucky :stuck_out_tongue:

laumann commented on src/lib.rs:
```diff
@@ -447,89 +447,68 @@ pub fn iselect<E, P, A>(chans: &Vec<Chan<E, Recv<A, P>>>) -> usize {
-        self.add_recv_ret(c, index);
-    }
+// /// Default use of ChanSelect works with usize and returns the index
+// /// of the selected channel. This is also the implementation used by
```
Just remove commented code.

laumann commented on src/lib.rs:
````diff
@@ -575,7 +554,7 @@ mod private {
 /// we can use the `offer!` macro as follows:
 ///
 /// ```rust
-/// #[macro_use] extern crate session_types;
````
I'm curious: Does this break compatibility with older versions of rustc?

laumann commented on tests/chan_select.rs:
```diff
@@ -62,13 +82,8 @@ fn chan_select_simple() {
 
-#[cfg(feature = "chan_select")]
 #[test]
 fn chan_select_add_ret() {
```
I'm kinda thinking this test could just be removed - the whole idea with the `_ret` stuff was to allow users to select the types returned from `sel.wait()` (in this test the `ChanToRead`). But that whole thing is gone now, so I'm not sure this test adds anything.

Munksgaard commented:
> I've just published `crossbeam-channel` version 0.3.3 with the new features so you can add it to `Cargo.toml`.

That's great news. Thank you @stjepang :+1:

Munksgaard commented:
@laumann What do you think, should we merge it? I think I've addressed all of your comments. 

I'm thinking that this probably necessitates a bump to 0.3.0 for us. While the outward API hasn't changed, the crate is suddenly fully usable on stable and you don't need to use the `chan_select` feature any more.

laumann commented on .travis.yml:
```diff
@@ -16,7 +16,7 @@ script:
 env:
   global:
-    - TRAVIS_CARGO_NIGHTLY_FEATURE=chan_select
+    - TRAVIS_CARGO_NIGHTLY_FEATURE=""
```
Can this be removed?

Munksgaard replied:
No, if we don't define `TRAVIS_CARGO_NIGHTLY_FEATURE` it is automatically set to `unstable`, which is not a feature in our library. See the docs here: https://github.com/huonw/travis-cargo

laumann commented on examples/hselect.rs:
```diff
@@ -0,0 +1,28 @@
+    println!("Got a response: \"{}\"", s);
+    to_close.close();
+
+    let _ = tcu.send("Hello, World from TCU!".to_string()).close();
```
I think all the call to `close()` do not need the `let _ = …`, ie

```
tcu.send("Hello, World from TCU!".to_string()).close();
```
should suffice.

(This applies in a few places.)

laumann commented on tests/chan_select.rs:
```diff
@@ -19,7 +14,32 @@ macro_rules! recv_assert_eq_close(
+
+    let receivers = vec!(rcs, rcu);
+
+    let () = tcs.send(1u64).close();
```
There's `let () = …` in a few places that I think could just be removed.

laumann commented on tests/tests.rs:
```diff
@@ -1,22 +1,17 @@
+    config.mode = mode.parse().expect("Invalid mode");
     config.src_base = PathBuf::from(format!("tests/{}", mode));
-    config.target_rustcflags = Some("-L target/debug".to_string());
+    config.link_deps();
```
Using `link_deps()` has a tendency to break on Windows, because of spaces in paths, so I'd be careful when using it :-)

Munksgaard replied:
The tests stop working on my machine if I replace `config.link_deps()` with `config.target_rustcflags = Some("-L target/debug".to_string());`. Any idea why?
